src/python/pants/util/collections.py

```diff
@@ -95,7 +95,15 @@ def partition_sequentially(
     # To stably partition the arguments into ranges of at least `size_min`, we sort them, and
     # create a new batch sequentially once we have the minimum number of entries, _and_ we encounter
     # an item hash prefixed with a threshold of zeros.
-    zero_prefix_threshold = math.log(size_min // 8, 2)
+    #
+    # The hashes act like a (deterministic) series of rolls of an evenly distributed die. The
+    # probability of a hash prefixed with Z zero bits is 1/2^Z, and so to break after N items on
+    # average, we look for `Z == log2(N)` zero bits.
+    #
+    # Breaking on these deterministic boundaries means that adding any single item will affect
```

Comment:
Thanks. Ok, but this assumes ... the same CLI specs with files edits / adds in between? In other words this whole scheme does ~nothing for `./pants lint bob::` followed by `./pants lint ::` - those will likely have completely different buckets - no hits at all. If that's right and all this fanciness is to support repeatedly running the same Pants command - that's great but seems useful to call out somewhere. I'm not sure where. Here is not the place since this would be about the larger picture and the call after call inputs to this function from a higher layer. If I've got this wrong and this always works - that's awesome. I'm still mystified though how that works. If I've got it right, it would be great to get some comment that does this spelling out that we're optimizing a particular use pattern.

Reply:
> Thanks. Ok, but this assumes ... the same CLI specs with files edits / adds in between? In other words this whole scheme does ~nothing for `./pants lint bob::` followed by `./pants lint ::` - those will likely have completely different buckets - no hits at all.

This is mostly about optimizing the re-running of a single command over time, yea. But because the inputs are sorted before hashing, `bob::` should fall entirely into sequential buckets within the entire set, and some of those might be able to hit in the larger set.

Unfortunately, it's hard to make promises about that in the presence of the `min`/`max` thresholds. If we were breaking _purely_ along bucket boundaries (without additionally setting minimum sizes), we could make better guarantees (because the bucket boundaries within `bob::` would be the same when the rest of `::` was added).

Thanks for raising this point though... it does actually seem like a desirable enough use case to try and strengthen this further.

Reply:
I'm not sure if strengthening this further, or even using the current trickery, belongs in this PR. IIUC batching is purely a win over the status quo today, even if the batches are simply formed from fixed size buckets over the sorted input and nothing more. If that's true, then this PR is only muddied by the fiddly bits here and maybe adding the fiddly bits en-masse could be done as a follow up that gives magic speed ups for free.

Reply:
The reason I think that doing it here might be useful is that it would adjust the meaning of the `--batch-size` flag (from "minimum size" to "target size")... it should be a quick change I think.

Reply:
@Eric-Arellano brought up a good point above about the fact that changing the batch size or implementation could have negative repercussions (due to the cache being invalidated). However as @jsirois points out any batching is a win.

IMO the solution with #9964 is the "baseline" and any improvement over that is gravy. If Pants finds better solutions/numbers over time I'm OK with a one-time cache bust which result in speedier runs going forward.

Because of the amount of variance here with tools/machine/repo I think the best approach is probably a data-driven one. You might try building a performance test suite to test various numbers/algorithms and asking the community (and/or Toolchain's customers) to run it overnight (or during a dead time) on a machine or two. I know I'd happily volunteer some CPU cycles to help find the optimal config.

Reply:
> The reason I think that doing it here might be useful is that it would adjust the meaning of the --batch-size flag (from "minimum size" to "target size")

Imo it's fine to change the behavior up till 2.10.0rc0. This seems worth landing as-is

Reply:
...it was not, in fact, a "quick change", because removing the min threshold and attempting to use _only_ the hash to choose boundaries requires increasing the threshold to the point where our sweet spot of bucket size (~128) is too likely to never encounter a good boundary.

Since the code as-is already snaps to boundaries, I feel fairly confident that we'll already get some matches in the `bob::` vs `::` case, but can defer looking at that to a followup.

Reply:
I couldn't help myself. #14210.